packages/analytics-core/src/plugins/destination.ts

```diff
@@ -78,87 +77,94 @@ export class Destination implements DestinationPlugin {
-    list.forEach((context) => {
-      // Only need to concat the queue for the first time
-      if (shouldAddToQueue) {
-        this.queue = this.queue.concat(context);
-      }
-      if (context.timeout === 0) {
-        this.schedule(this.config.flushIntervalMillis);
-        return;
+  increaseAttempts(list: Context[]) {
+    const updateEventSet = new Set(list.map((context) => context.event.insert_id));
+    this.queue.forEach((context) => {
+      if (updateEventSet.has(context.event.insert_id)) {
+        context.attempts += 1;
       }
-
-      setTimeout(() => {
-        context.timeout = 0;
-        this.schedule(0);
-      }, context.timeout);
     });
   }
 
-  addToQueue(...list: Context[]) {
-    const tryable = this.getTryableList(list);
-    this.scheduleTryable(tryable, true);
-    this.saveEvents();
-  }
+  sendEventsIfReady() {
+    if (this.config.offline) {
+      return;
+    }
 
-  schedule(timeout: number) {
-    if (this.scheduled || this.config.offline) {
+    if (this.queue.length >= this.config.flushQueueSize) {
+      void this.flush(true);
+    }
+
+    if (this.scheduled) {
       return;
     }
 
     this.scheduled = setTimeout(() => {
       void this.flush(true).then(() => {
         if (this.queue.length > 0) {
-          this.schedule(timeout);
+          this.sendEventsIfReady();
         }
       });
-    }, timeout);
+    }, this.config.flushIntervalMillis);
+
+    return;
   }
 
+  // flush the queue
   async flush(useRetry = false) {
     // Skip flush if offline
     if (this.config.offline) {
       this.config.loggerProvider.debug('Skipping flush while offline.');
       return;
     }
 
-    const list: Context[] = [];
-    const later: Context[] = [];
-    this.queue.forEach((context) => (context.timeout === 0 ? list.push(context) : later.push(context)));
-
     if (this.scheduled) {
       clearTimeout(this.scheduled);
       this.scheduled = null;
     }
 
-    const batches = chunk(list, this.config.flushQueueSize);
-
-    await Promise.all(batches.map((batch) => this.send(batch, useRetry)));
+    this.filterTriableList();
+    if (this.queue.length === 0) {
+      return;
+    }
 
-    this.scheduleTryable(later);
+    const batches = chunk(this.queue, this.config.flushQueueSize);
+    for (const batch of batches) {
+      this.increaseAttempts(batch);
+      const responseStatus = await this.send(batch, useRetry);
+      // To keep the order of events sending to the backend, stop sending the other requests while error happending.
+      // All the events are still queued, and will be retried with next trigger.
+      if (retryStatus.has(responseStatus) && useRetry) {
+        break;
+      }
+    }
   }
 
   async send(list: Context[], useRetry = true) {
     if (!this.config.apiKey) {
-      return this.fulfillRequest(list, 400, MISSING_API_KEY_MESSAGE);
+      this.fulfillRequest(list, 400, MISSING_API_KEY_MESSAGE);
+      return Status.Invalid;
```

Comment:
Should we add the return type to the function's signature too?

Reply:
Added